File: app/src/androidTest/java/com/cottondroid/todoachieved/task/list/TaskListActivityTest.java
package com.cottondroid.todoachieved.task.list;

import android.app.Activity;
import android.app.Instrumentation;
import android.content.Intent;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.espresso.intent.Intents;
import androidx.test.rule.ActivityTestRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.cottondroid.todoachieved.R;
import com.cottondroid.todoachieved.TodoApplication;
import com.cottondroid.todoachieved.di.DaggerTestApplicationComponent;
import com.cottondroid.todoachieved.di.TestApplicationComponent;
import com.cottondroid.todoachieved.matchers.RecyclerViewMatcher;
import com.cottondroid.todoachieved.task.form.TaskFormActivity;
import com.cottondroid.todoachieved.task.model.TodoTask;
import com.cottondroid.todoachieved.task.model.TodoTaskDao;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.intent.Intents.intended;
import static androidx.test.espresso.intent.Intents.intending;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent;
import static androidx.test.espresso.matcher.ViewMatchers.*;
import static org.hamcrest.Matchers.not;

@RunWith(AndroidJUnit4.class)
public class TaskListActivityTest {

    private static final int TASK_COUNT = 10;
    private static final String TASK_PREFIX = "Task ";

    @Inject
    TodoTaskDao todoTaskDao;

    @Rule
    public ActivityTestRule<TaskListActivity> activityTestRule = new ActivityTestRule<>(TaskListActivity.class, false, false);

    @Before
    public void setUp() {
        TodoApplication application = (TodoApplication) InstrumentationRegistry.getInstrumentation().getTargetContext().getApplicationContext();
        TestApplicationComponent component = DaggerTestApplicationComponent
                .builder()
                .application(application)
                .build();
        component.inject(application);
        component.inject(this);
        addTasks(TASK_COUNT);
        Intents.init();
        intending(not(hasComponent(TaskListActivity.class.getCanonicalName()))).respondWith(new Instrumentation.ActivityResult(Activity.RESULT_OK, null));
    }

    private List<TodoTask> addTasks(int taskCount) {
        List<TodoTask> sampleTasks = new ArrayList<>();
        for (int i = 0; i < taskCount; i++) {
            todoTaskDao.insertOrReplace(new TodoTask((long) i, TASK_PREFIX + i, new Date().getTime(), new Date().getTime()));
        }
        return sampleTasks;
    }

    @After
    public void tearDown() {
        Intents.release();
    }

    @Test
    public void showItems() {

        activityTestRule.launchActivity(new Intent());
        for (int i = 0; i < TASK_COUNT; i++) {
            onView(new RecyclerViewMatcher((R.id.recycler_view)).atPosition(i))
                    .check(matches(hasDescendant(withText(TASK_PREFIX + i))));
        }
    }

    @Test
    public void addItem() {

        activityTestRule.launchActivity(new Intent());

        onView(withId(R.id.floating_action_button)).perform(click());

        intended(hasComponent(TaskFormActivity.class.getCanonicalName()));

    }
}

File: app/src/androidTest/java/com/cottondroid/todoachieved/task/list/TaskListActivityTest.kt
package com.cottondroid.todoachieved.task.list

import android.app.Activity
import android.app.Instrumentation.ActivityResult
import android.content.Intent
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.intent.Intents
import androidx.test.espresso.intent.matcher.IntentMatchers
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import com.cottondroid.todoachieved.R
import com.cottondroid.todoachieved.TodoApplication
import com.cottondroid.todoachieved.di.DaggerTestApplicationComponent
import com.cottondroid.todoachieved.di.TestApplicationComponent
import com.cottondroid.todoachieved.matchers.RecyclerViewMatcher
import com.cottondroid.todoachieved.rule.RxSchedulerRule
import com.cottondroid.todoachieved.task.form.TaskFormActivity
import com.cottondroid.todoachieved.task.model.TodoTask
import com.cottondroid.todoachieved.task.model.TodoTaskDao
import org.hamcrest.Matchers
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import java.util.ArrayList
import java.util.Date
import javax.inject.Inject

class TaskListActivityTest {
    @Inject
    lateinit var todoTaskDao: TodoTaskDao

    @Rule
    @JvmField
    val activityTestRule = ActivityTestRule(TaskListActivity::class.java, false, false)

    @Rule
    @JvmField
    val rxSchedulerRule = RxSchedulerRule()

    @Before
    fun setUp() {
        val application = InstrumentationRegistry.getInstrumentation().targetContext.applicationContext as TodoApplication
        val component: TestApplicationComponent = DaggerTestApplicationComponent
                .builder()
                .application(application)
                .build()
        component.inject(application)
        component.inject(this)
        addTasks()
        Intents.init()
        Intents.intending(Matchers.not(IntentMatchers.hasComponent(TaskListActivity::class.java.canonicalName)))
                .respondWith(ActivityResult(Activity.RESULT_OK, null))
    }

    private fun addTasks(): List<TodoTask> {
        val sampleTasks: List<TodoTask> = ArrayList()
        for (i in 0 until TASK_COUNT) {
            todoTaskDao.insertOrReplace(
                    TodoTask(
                            id = i.toLong(),
                            serverId = i.toString(),
                            text = TASK_PREFIX + i,
                            createdDate = Date().time
                    )
            ).subscribe()
        }
        return sampleTasks
    }

    @After
    fun tearDown() {
        Intents.release()
    }

    @Test
    fun showItems() {
        activityTestRule.launchActivity(Intent())
        for (i in 0 until TASK_COUNT) {
            Espresso.onView(RecyclerViewMatcher(R.id.recyclerView).atPosition(i))
                    .check(ViewAssertions.matches(ViewMatchers.hasDescendant(ViewMatchers.withText(TASK_PREFIX + i))))
        }
    }

    @Test
    fun addItem() {
        activityTestRule.launchActivity(Intent())
        Espresso.onView(withId(R.id.fab)).perform(ViewActions.click())
        Intents.intended(IntentMatchers.hasComponent(TaskFormActivity::class.java.canonicalName))
    }

    companion object {
        private const val TASK_COUNT = 10
        private const val TASK_PREFIX = "Task "
    }
}